lib/puppet/pops/lookup/context.rb

```diff
@@ -0,0 +1,81 @@
+module Puppet::Pops
+  module Lookup
+    class Context
+      include Types::PuppetObject
+
+      def self._ptype
+        @type
+      end
+
+      def self.register_ptype(loader, ir)
+        tf = Types::TypeFactory
+        @type = Pcore::create_object_type(loader, ir, self, 'Puppet::LookupContext', 'Any',
+          {
+            'environment_name' => Types::PStringType::NON_EMPTY,
+            'module_name' => {
+              Types::KEY_TYPE => tf.optional(Types::PStringType::NON_EMPTY),
+              Types::KEY_VALUE => nil
+            }
+          },
+          {
+            'not_found' => tf.callable([0, 0], tf.undef),
+            'explain' => tf.callable([0, 0, tf.callable(0,0)], tf.undef),
+            'cache' => tf.callable([tf.scalar, tf.any], tf.undef),
+            'cache_all' => tf.callable([tf.hash_kv(tf.scalar, tf.any)], tf.undef),
+            'cached_value' => tf.callable([tf.scalar], tf.any),
+            'cached_entries' => tf.variant(
+              tf.callable([0, 0, tf.callable(2,2)], tf.undef),
```

Comment:
If it is to mimic each, then also support 1 arg in the lambda to get each key/value tuple ?